File: Python/Algorithms/Tree/AVLTree.py
"""
Before going through the code, have a look at the following blog about AVL Tree:
https://en.wikipedia.org/wiki/AVL_tree

@author: lashuk1729

"""
class Node():
    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None
        self.height = 1

class AVL_Tree():
    def get_height(self, root):
        if root is None:
            return 0
        return root.height
    
    def get_balance(self, root):
        """
        Checks whether the given tree is balanced or not, i.e., 
        difference between height of left_subtree and right_subtree should lie between -1 & 1
        """
        if root is None:
            return 0
        return self.get_height(root.left) - self.get_height(root.right)
    
    def inorder(self, root):
        if root:
            self.inorder(root.left)
            print(root.val, end = " -> ")
            self.inorder(root.right)
    
    def preorder(self, root):
        if root:
            print(root.val, end = " -> ")
            self.preorder(root.left)
            self.preorder(root.right)
    
    def right_rotate(self, z):
        y = z.left
        T3 = y.right
        y.right = z
        z.left = T3
        z.height = 1 + max(self.get_height(z.left), self.get_height(z.right))
        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))
        return y
    
    def left_rotate(self, z):
        y = z.right
        T3 = y.left
        y.left = z
        z.right = T3
        z.height = 1 + max(self.get_height(z.left), self.get_height(z.right))
        y.height = 1 + max(self.get_height(y.left), self.get_height(y.right))
        return y
    
    def get_min(self, root):
        while root.left is not None:
            root = root.left
        return root
    
    def insert(self, root, key):
        if root is None:
            return Node(key)
        elif root.val > key:
            root.left = self.insert(root.left, key)
        else:
            root.right = self.insert(root.right, key)
        
        root.height = 1 + max(self.get_height(root.left), self.get_height(root.right))
        balance = self.get_balance(root)
        # Left-left case
        if balance > 1 and key < root.left.val:
            return self.right_rotate(root)
        # Right-right case
        if balance < -1 and key > root.right.val:
            return self.left_rotate(root)
        # Left-right case
        if balance > 1 and key > root.left.val:
            root.left = self.left_rotate(root.left)
            return self.right_rotate(root)
        # Right-left case
        if balance < -1 and key < root.right.val:
            root.right = self.right_rotate(root.right)
            return self.left_rotate(root)
        return root
    
    def delete(self, root, key):
        if root is None:
            return root
        elif root.val > key:
            root.left = self.delete(root.left, key)
        elif root.val < key:
            root.right = self.delete(root.right, key)
        else:
            if root.left is None:
                temp = root.right
                root = None
                return temp
            elif root.right is None:
                temp = root.left
                root = None
                return temp
            temp = self.get_min(root.right)
            root.key = temp.key
            root.right = self.delete(root.right, temp.key)
            
        if root is None:
            return root
        root.height = 1 + max(self.get_height(root.left), self.get_height(root.right))
        balance = self.get_balance(root)
        # Left-left case
        if balance > 1 and self.get_balance(root.left) >= 0:
            return self.right_rotate(root)
        # Right-right case
        if balance < -1 and self.get_balance(root.right) <= 0:
            return self.left_rotate
        # Left-right case
        if balance > 1 and self.get_balance(root.left) < 0:
            root.left = self.left_rotate(root.left)
            return self.right_rotate(root)
        # Right-left case
        if balance < -1 and self.get_balance(root.right) > 0:
            root.right = self.right_rotate(root.right)
            return self.left_rotate(root)
        return root
    
if __name__=='__main__':
    myTree = AVL_Tree() 
    root = None

    nums = [9, 5, 10, 0, 6, 11, -1, 1, 2]

    for num in nums:
        root = myTree.insert(root, num)

    # Preorder Traversal
    print("AVL tree after insertion is: ", end = "") 
    myTree.preorder(root) 
    print('X')
    
    key = 10
    root = myTree.delete(root, key)
    
    # Preorder Traversal
    print("AVL tree after deletion is: ", end = "")
    myTree.preorder(root)
    print('X')

File: Javascript/Data-structures/Tree/avlTree.js
// Create node
const Node = function (item) {
  this.item = item;
  this.height = 1;
  this.left = null;
  this.right = null;
};

//AVL Tree
const AVLTree = function () {
  let root = null;

  //return height of the node
  this.height = (N) => {
    if (N === null) {
      return 0;
    }

    return N.height;
  };

  //right rotate
  this.rightRotate = (y) => {
    let x = y.left;
    let T2 = x.right;
    x.right = y;
    y.left = T2;
    y.height = Math.max(this.height(y.left), this.height(y.right)) + 1;
    x.height = Math.max(this.height(x.left), this.height(x.right)) + 1;
    return x;
  };

  //left rotate
  this.leftRotate = (x) => {
    let y = x.right;
    let T2 = y.left;
    y.left = x;
    x.right = T2;
    x.height = Math.max(this.height(x.left), this.height(x.right)) + 1;
    y.height = Math.max(this.height(y.left), this.height(y.right)) + 1;
    return y;
  };

  // get balance factor of a node
  this.getBalanceFactor = (N) => {
    if (N == null) {
      return 0;
    }

    return this.height(N.left) - this.height(N.right);
  };

  // helper function to insert a node
  const insertNodeHelper = (node, item) => {
    // find the position and insert the node
    if (node === null) {
      return new Node(item);
    }

    if (item < node.item) {
      node.left = insertNodeHelper(node.left, item);
    } else if (item > node.item) {
      node.right = insertNodeHelper(node.right, item);
    } else {
      return node;
    }

    // update the balance factor of each node
    // and, balance the tree
    node.height = 1 + Math.max(this.height(node.left), this.height(node.right));

    let balanceFactor = this.getBalanceFactor(node);

    if (balanceFactor > 1) {
      if (item < node.left.item) {
        return this.rightRotate(node);
      } else if (item > node.left.item) {
        node.left = this.leftRotate(node.left);
        return this.rightRotate(node);
      }
    }

    if (balanceFactor < -1) {
      if (item > node.right.item) {
        return this.leftRotate(node);
      } else if (item < node.right.item) {
        node.right = this.rightRotate(node.right);
        return this.leftRotate(node);
      }
    }

    return node;
  };

  // insert a node
  this.insertNode = (item) => {
    // console.log(root);
    root = insertNodeHelper(root, item);
  };

  //get node with minimum value
  this.nodeWithMimumValue = (node) => {
    let current = node;
    while (current.left !== null) {
      current = current.left;
    }
    return current;
  };

  // delete helper
  const deleteNodeHelper = (root, item) => {
    // find the node to be deleted and remove it
    if (root == null) {
      return root;
    }
    if (item < root.item) {
      root.left = deleteNodeHelper(root.left, item);
    } else if (item > root.item) {
      root.right = deleteNodeHelper(root.right, item);
    } else {
      if (root.left === null || root.right === null) {
        let temp = null;
        if (temp == root.left) {
          temp = root.right;
        } else {
          temp = root.left;
        }

        if (temp == null) {
          temp = root;
          root = null;
        } else {
          root = temp;
        }
      } else {
        let temp = this.nodeWithMimumValue(root.right);
        root.item = temp.item;
        root.right = deleteNodeHelper(root.right, temp.item);
      }
    }
    if (root == null) {
      return root;
    }

    // Update the balance factor of each node and balance the tree
    root.height = Math.max(this.height(root.left), this.height(root.right)) + 1;

    let balanceFactor = this.getBalanceFactor(root);
    if (balanceFactor > 1) {
      if (this.getBalanceFactor(root.left) >= 0) {
        return this.rightRotate(root);
      } else {
        root.left = this.leftRotate(root.left);
        return this.rightRotate(root);
      }
    }
    if (balanceFactor < -1) {
      if (this.getBalanceFactor(root.right) <= 0) {
        return this.leftRotate(root);
      } else {
        root.right = this.rightRotate(root.right);
        return this.leftRotate(root);
      }
    }
    return root;
  };

  //delete a node
  this.deleteNode = (item) => {
    root = deleteNodeHelper(root, item);
  };

  // print the tree in pre - order
  this.preOrder = () => {
    preOrderHelper(root);
  };

  const preOrderHelper = (node) => {
    if (node) {
      console.log(node.item);
      preOrderHelper(node.left);
      preOrderHelper(node.right);
    }
  };
};

//   Insert:
var tree = new AVLTree();
tree.insertNode(33);
tree.insertNode(13);
tree.insertNode(53);
tree.insertNode(9);
tree.insertNode(21);
tree.insertNode(61);
tree.insertNode(8);
tree.insertNode(11);
tree.preOrder();
tree.deleteNode(13);
console.log("After Deletion: ");
tree.preOrder();
